### PR TITLE
Write Unit Tests for PredictionsApi, Add more Documentations and Code cleanups

### LINKED ISSUE #18: add unit tests for PredictionsApi

### PR DESCRIPTION
**Pull request by mayorjay**
Closes #18

### CODE REVIEW

File: src/test/kotlin/io/github/enyason/base/RetrofitFactoryTest.kt [hunk 1]
```diff
@@ -2,39 +2,37 @@ package io.github.enyason.base
...
 import kotlin.test.assertNotNull
 import kotlin.test.assertTrue
 
-@TestMethodOrder(MethodOrderer.OrderAnnotation::class)
 class RetrofitFactoryTest {
 
+    @AfterTest
+    fun tearDown() {
+        RetrofitFactory.reset()
+    }
```
Comment (mayorjay) [inline src/test/kotlin/io/github/enyason/base/RetrofitFactoryTest.kt:18]: Due to the fact that RetrofitFactory is a singleton, the tests in this class were affecting the tests in PredictionsApiTest. Hence the need for making Retrofit nullable and creating the reset function

File: src/main/kotlin/io/github/enyason/base/RetrofitFactory.kt [hunk 2]
```diff
@@ -13,44 +14,48 @@ import retrofit2.converter.gson.GsonConverterFactory
...
+     * @param config A set of configurations used to control the way the SDK behaves
      */
     fun buildRetrofit(
-        config: ReplicateConfig,
-        loggingLevel: HttpLoggingInterceptor.Level = HttpLoggingInterceptor.Level.BASIC
+        config: ReplicateConfig
     ): Retrofit {
-        if (this::retrofit.isInitialized) {
-            return retrofit
+        if (retrofit != null) {
```
Comment (enyadobande) [inline src/main/kotlin/io/github/enyason/base/RetrofitFactory.kt:28]: a little nit-picky, but we can assign the retrofit instance to a local retrofit?
Using a local variable we can avoid using the double bang after the null check.

File: src/main/kotlin/io/github/enyason/base/RetrofitFactory.kt [hunk 3]
```diff
@@ -13,44 +14,48 @@ import retrofit2.converter.gson.GsonConverterFactory
...
         }
 
         retrofit = Retrofit.Builder()
             .baseUrl(config.baseUrl)
-            .client(buildHttpClient(config, loggingLevel))
+            .client(buildHttpClient(config))
             .addConverterFactory(createConverterFactory())
             .build()
-        return retrofit
+        return retrofit!!
```
Comment (enyadobande) [inline src/main/kotlin/io/github/enyason/base/RetrofitFactory.kt]: I just dislike double bang 😀

File: src/test/kotlin/io/github/enyason/predictions/PredictionsApiTest.kt [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+    private lateinit var mockWebServer: MockWebServer
+    private lateinit var sut: PredictionsApi
+    private val gson = Gson()
+
+    @BeforeTest
+    fun setup() {
+        mockWebServer = MockWebServer()
+        mockWebServer.start(8080)
+        sut = PredictionsApi(ReplicateConfig("tokenYUIOPJHIGUFYDRT", baseUrl = mockWebServer.url("").toString()))
```
Comment (enyadobande) [inline src/test/kotlin/io/github/enyason/predictions/PredictionsApiTest.kt]: why is the baseUrl set from the mockWebServer.url with an empty string?

File: src/test/kotlin/io/github/enyason/predictions/PredictionsApiTest.kt [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    @AfterTest
+    fun tearDown() {
+        mockWebServer.shutdown()
+    }
+
+    @Test
+    fun `test createPrediction _API returns success response`() = runTest {
+        val predictable = TestPredictable(versionId = "2exbc4", input = mapOf("prompt" to "hd image of Einstein"))
```
Comment (enyadobande) [inline src/test/kotlin/io/github/enyason/predictions/PredictionsApiTest.kt]: can we extract the input into a variable so you can use it on line 41?

**Review by enyadobande — APPROVED** [review]
Pretty good stuff. Well done

File: src/test/kotlin/io/github/enyason/predictions/PredictionsApiTest.kt (hunk 4 above)
Comment (mayorjay) [reply src/test/kotlin/io/github/enyason/predictions/PredictionsApiTest.kt]: The url functions accepts a path which should either be an empty string or **/**

File: src/main/kotlin/io/github/enyason/base/RetrofitFactory.kt (hunk 2 above)
Comment (mayorjay) [reply src/main/kotlin/io/github/enyason/base/RetrofitFactory.kt:28]: I'm not sure I get what you mean, maybe an example will be helpful

Comment (enyason) [reply src/main/kotlin/io/github/enyason/base/RetrofitFactory.kt:28]: here you go:


 ```
 fun buildRetrofit(
        config: ReplicateConfig
    ): Retrofit {

        val retrofit = this.retrofit

        if (retrofit != null) {
            return retrofit
        }

        return Retrofit.Builder()
            .baseUrl(config.baseUrl)
            .client(buildHttpClient(config))
            .addConverterFactory(createConverterFactory())
            .build().also { instance ->
                this.retrofit = instance
            }
    }

```

@mayorJAY

Comment (enyason) [reply src/main/kotlin/io/github/enyason/base/RetrofitFactory.kt:28]: see example above

**Review by enyason — CHANGES_REQUESTED** [review]
check my example

File: src/main/kotlin/io/github/enyason/base/RetrofitFactory.kt (hunk 2 above)
Comment (mayorjay) [reply src/main/kotlin/io/github/enyason/base/RetrofitFactory.kt:28]: Oh, I see. Thanks

**Pull request merged**